# JOBSHEET 13 - Artificial Neural Network (ANN) dan Evaluasi Classifier

Nama: Diana Rahmawati <br>
Kelas: TI - 3G <br>
NIM: 2341720162

### PRAKTIKUM 1

In [ ]:
import numpy as np

# Dataset XOR
X = np.array([[0,0],[0,1],[1,0],[1,1]])
y = np.array([[0],[1],[1],[0]])

# Parameter
input_size = 2
hidden_size = 2
output_size = 1
lr = 0.1

# Inisialisasi bobot
W1 = np.random.randn(input_size, hidden_size)
b1 = np.zeros((1, hidden_size))
W2 = np.random.randn(hidden_size, output_size)
b2 = np.zeros((1, output_size))

# Fungsi aktivasi
def sigmoid(x):
    return 1 / (1 + np.exp(-x))

def sigmoid_derivative(x):
    return x * (1 - x)

# Training
for epoch in range(10000):
    # Forward pass
    z1 = np.dot(X, W1) + b1
    a1 = sigmoid(z1)
    z2 = np.dot(a1, W2) + b2
    a2 = sigmoid(z2)

    # Hitung error
    error = y - a2

    # Backpropagation
    d_a2 = error * sigmoid_derivative(a2)
    d_W2 = np.dot(a1.T, d_a2)
    d_b2 = np.sum(d_a2, axis=0, keepdims=True)

    d_a1 = np.dot(d_a2, W2.T) * sigmoid_derivative(a1)
    d_W1 = np.dot(X.T, d_a1)
    d_b1 = np.sum(d_a1, axis=0, keepdims=True)

    # Update bobot
    W1 += lr * d_W1
    b1 += lr * d_b1
    W2 += lr * d_W2
    b2 += lr * d_b2

    if epoch % 1000 == 0:
        loss = np.mean(np.square(error))
        print(f"Epoch {epoch}, Loss: {loss}")

# Output akhir
print("Prediksi:")
print(a2)

Epoch 0, Loss: 0.35031570609791385
Epoch 1000, Loss: 0.25017406939631287
Epoch 2000, Loss: 0.2500205007952759
Epoch 3000, Loss: 0.24990855508944518
Epoch 4000, Loss: 0.24965786651841557
Epoch 5000, Loss: 0.2485338814808645
Epoch 6000, Loss: 0.23983285075708838
Epoch 7000, Loss: 0.20132145453223627
Epoch 8000, Loss: 0.1290264330672145
Epoch 9000, Loss: 0.040523563096821785
Prediksi:
[[0.13598044]
 [0.85711266]
 [0.84587071]
 [0.12002351]]


### Tugas 1

In [ ]:
import numpy as np

# Dataset XOR
X = np.array([[0,0],[0,1],[1,0],[1,1]])
y = np.array([[0],[1],[1],[0]])

# Parameter
input_size = 2
hidden_size = 3
output_size = 1
lr = 0.1

# Inisialisasi bobot
W1 = np.random.randn(input_size, hidden_size)
b1 = np.zeros((1, hidden_size))
W2 = np.random.randn(hidden_size, output_size)
b2 = np.zeros((1, output_size))

# Fungsi aktivasi
def sigmoid(x):
    return 1 / (1 + np.exp(-x))

def sigmoid_derivative(x):
    return x * (1 - x)

# Training
for epoch in range(10000):
    # Forward pass
    z1 = np.dot(X, W1) + b1
    a1 = sigmoid(z1)
    z2 = np.dot(a1, W2) + b2
    a2 = sigmoid(z2)

    # Hitung error
    error = y - a2

    # Backpropagation
    d_a2 = error * sigmoid_derivative(a2)
    d_W2 = np.dot(a1.T, d_a2)
    d_b2 = np.sum(d_a2, axis=0, keepdims=True)

    d_a1 = np.dot(d_a2, W2.T) * sigmoid_derivative(a1)
    d_W1 = np.dot(X.T, d_a1)
    d_b1 = np.sum(d_a1, axis=0, keepdims=True)

    # Update bobot
    W1 += lr * d_W1
    b1 += lr * d_b1
    W2 += lr * d_W2
    b2 += lr * d_b2

    if epoch % 1000 == 0:
        loss = np.mean(np.square(error))
        print(f"Epoch {epoch}, Loss: {loss}")

# Output akhir
print("Prediksi:")
print(a2)

Epoch 0, Loss: 0.46085569994820663
Epoch 1000, Loss: 0.21202246573158828
Epoch 2000, Loss: 0.07841556799504917
Epoch 3000, Loss: 0.02522420746598441
Epoch 4000, Loss: 0.012724704480768807
Epoch 5000, Loss: 0.0081165034028928
Epoch 6000, Loss: 0.0058466389318458075
Epoch 7000, Loss: 0.0045256933336019324
Epoch 8000, Loss: 0.0036715316664129097
Epoch 9000, Loss: 0.0030779834002026274
Prediksi:
[[0.02161483]
 [0.94754139]
 [0.94534069]
 [0.06609632]]


Perbandingan hasil loss pada saat menggunakan neuron hidden layer 2 dan 3:
- Neuron 2: Loss awal: 0.3503, Loss akhir (epoch 9000): 0.0405
Loss turun, tetapi konvergensi lambat dan relatif besar untuk XOR. Model butuh waktu lama untuk mendekati pola XOR
- Neuron 3: Loss awal: 0.4608 (lebih besar karena bobot acak), Loss akhir (epoch 9000): 0.0030. Dan terlihat jauh lebih cepat turun (epoch 2000 sudah 0.078, sedangkan model 2 neuron butuh 8000 epoch untuk turun drastis), jauh lebih kecil (dari 0.040 → 0.003), dan lebih stabil.


In [ ]:
import numpy as np

# --- Activation ---
def sigmoid(x):
    return 1 / (1 + np.exp(-x))

def sigmoid_deriv(x):
    s = sigmoid(x)
    return s * (1 - s)

def relu(x):
    return np.maximum(0, x)

def relu_derivative(x):
    return (x > 0).astype(float)

# --- Dataset contoh ---
X = np.array([[0,0],[0,1],[1,0],[1,1]])
y = np.array([[0],[1],[1],[0]])

# --- Parameter ---
np.random.seed(0)
W1 = np.random.randn(2, 4) * 0.1
b1 = np.zeros((1, 4))
W2 = np.random.randn(4, 1) * 0.1
b2 = np.zeros((1, 1))

lr = 0.1

# --- Training ---
for epoch in range(10000):

    # Forward
    z1 = X.dot(W1) + b1
    a1 = relu(z1)

    z2 = a1.dot(W2) + b2
    y_pred = sigmoid(z2)

    # Loss
    loss = np.mean((y_pred - y)**2)

    # Backprop
    dz2 = 2 * (y_pred - y) * sigmoid_deriv(z2)
    dW2 = a1.T.dot(dz2)
    db2 = np.sum(dz2, axis=0, keepdims=True)

    da1 = dz2.dot(W2.T)
    dz1 = da1 * relu_derivative(z1)
    dW1 = X.T.dot(dz1)
    db1 = np.sum(dz1, axis=0, keepdims=True)

    # Update
    W1 -= lr * dW1
    b1 -= lr * db1
    W2 -= lr * dW2
    b2 -= lr * db2

    if epoch % 1000 == 0:
        print(f"Epoch {epoch}, Loss: {loss}")

# --- Result ---
print("Prediksi akhir:")
print(y_pred)


Epoch 0, Loss: 0.24979120086251835
Epoch 1000, Loss: 0.0032663457441631195
Epoch 2000, Loss: 0.0009340787675389395
Epoch 3000, Loss: 0.0005222133132021824
Epoch 4000, Loss: 0.0003575487771457564
Epoch 5000, Loss: 0.00027012982889811874
Epoch 6000, Loss: 0.00021619774699801222
Epoch 7000, Loss: 0.0001798237753878717
Epoch 8000, Loss: 0.00015364740461270738
Epoch 9000, Loss: 0.00013398013929530442
Prediksi akhir:
[[0.01869746]
 [0.99237571]
 [0.99418944]
 [0.00576233]]


Perbandingan hasil :
- Menambah neuron (2 menjadi 3) dapat meningkatkan kemampuan model belajar berupa fungsi XOR yang non-linear. Mengganti Sigmoid menjadi ReLU memberi dampak paling besar. Gradien tidak mengecil, Jauh lebih cepat konvergen, Loss paling kecil, Prediksi paling akurat